File: app/src/main/java/com/group_finity/mascot/script/ScriptEngineManager.java
package com.group_finity.mascot.script;

import org.graalvm.polyglot.Context;
import org.graalvm.polyglot.Engine;
import org.graalvm.polyglot.HostAccess;
import org.graalvm.polyglot.Source;
import org.graalvm.polyglot.Value;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;

/**
 * GraalJS スクリプトエンジンの管理クラス。
 * 
 * <p>アーキテクチャ:
 * <ul>
 *   <li><b>Shared Engine:</b> JITコンパイル結果やコードキャッシュを共有し、メモリ効率とパフォーマンスを向上させます。</li>
 *   <li><b>Separate Contexts:</b> マスコット個体ごとに Context を分離し、変数のスコープを独立させます。</li>
 *   <li><b>Host Access:</b> Java の特定のクラスやメソッドへのアクセスを許可します。</li>
 * </ul>
 * </p>
 */
public class ScriptEngineManager {

    private static ScriptEngineManager instance;

    // 全インスタンスで共有するエンジン (JIT最適化の共有)
    private final Engine sharedEngine;

    // Java側からJSへのアクセス許可設定
    private final HostAccess hostAccess;

    private ScriptEngineManager() {
        // エンジンの初期化
        this.sharedEngine = Engine.newBuilder("js")
                .option("engine.WarnInterpreterOnly", "false") // 開発中の警告抑制
                .build();

        // ホストアクセスの設定 (アノテーション @HostAccess.Export があるもののみ許可)
        this.hostAccess = HostAccess.newBuilder(HostAccess.EXPLICIT)
                .allowListAccess(true) // 配列やリストへのアクセスを許可
                .allowMapAccess(true)  // Mapへのアクセスを許可 (JSオブジェクトのように扱えるようにする)
                .build();
    }

    public static synchronized ScriptEngineManager getInstance() {
        if (instance == null) {
            instance = new ScriptEngineManager();
        }
        return instance;
    }

    /**
     * マスコット個体用の新しいスクリプトコンテキストを作成します。
     * 
     * @param globalVariables コンテキスト初期化時に注入するグローバル変数 (例: "mascot" -> MascotInstance)
     * @return 初期化された Context
     */
    public Context createMascotContext(Map<String, Object> globalVariables) {
        Context context = Context.newBuilder("js")
                .engine(sharedEngine)
                .allowHostAccess(hostAccess)
                .allowHostClassLookup(className -> {
                    // 許可するJavaクラスのパッケージを制限 (必要に応じて調整)
                    return className.startsWith("com.group_finity.mascot.") ||
                           className.startsWith("java.awt.Point") ||
                           className.startsWith("java.awt.Rectangle");
                })
                .build();

        // グローバル変数の注入
        Value bindings = context.getBindings("js");
        if (globalVariables != null) {
            for (Map.Entry<String, Object> entry : globalVariables.entrySet()) {
                bindings.putMember(entry.getKey(), entry.getValue());
            }
        }

        return context;
    }

    /**
     * ファイルからスクリプトを読み込んでコンパイルします。
     * 結果の Source オブジェクトはキャッシュ可能です。
     */
    public Source loadScript(Path path) throws IOException {
        String content = Files.readString(path, StandardCharsets.UTF_8);
        return Source.newBuilder("js", content, path.getFileName().toString())
                .mimeType("application/javascript")
                .build();
    }

    /**
     * リソースの解放
     */
    public void shutdown() {
        if (sharedEngine != null) {
            sharedEngine.close();
        }
    }
}

File: app/src/main/kotlin/com/group_finity/mascot/script/ScriptEngineManager.kt
package com.group_finity.mascot.script

import org.graalvm.polyglot.Engine

/**
 * GraalJS Engine を管理するシングルトンオブジェクト。 複数の ScriptEngine (Context) で Engine を共有することで、
 * JITコンパイル結果やコードキャッシュを共有し、メモリ効率とパフォーマンスを向上させる。
 */
object ScriptEngineManager {
    private val engine: Engine =
            Engine.newBuilder("js")
                    // インタプリタモード実行時の警告を抑制
                    .option("engine.WarnInterpreterOnly", "false")
                    .build()

    /** 共有エンジンを使用して新しい ScriptEngine インスタンスを作成する。 */
    fun createEngine(): ScriptEngine {
        return ScriptEngine(engine)
    }

    /** 変数を初期化した状態で ScriptEngine を作成する。 */
    fun createMascotContext(variables: Map<String, Any>): ScriptEngine {
        val scriptEngine = createEngine()
        variables.forEach { (key, value) -> scriptEngine.put(key, value) }
        return scriptEngine
    }

    fun close() {
        engine.close()
    }

    /** ファイルからスクリプトを読み込んでコンパイルする。 */
    fun loadScript(path: java.nio.file.Path): org.graalvm.polyglot.Source {
        val content = java.nio.file.Files.readString(path, java.nio.charset.StandardCharsets.UTF_8)
        return org.graalvm.polyglot.Source.newBuilder("js", content, path.fileName.toString())
                .mimeType("application/javascript")
                .build()
    }

    /** 文字列コンテンツからスクリプトをコンパイルする。 */
    fun loadScript(content: String, name: String): org.graalvm.polyglot.Source {
        return org.graalvm.polyglot.Source.newBuilder("js", content, name)
                .mimeType("application/javascript")
                .build()
    }
}
